File: lib/src/models/characteristic_description.dart
import 'dart:typed_data';

class CharacteristicDescription {
  final String uuid;
  final Uint8List value;
  final bool read;
  final bool write;
  final bool writeNR; //Write no response
  final bool notify;
  final bool indicate;

  CharacteristicDescription({
    required String uuid,
    Uint8List? value,
    this.read = false,
    this.write = false,
    this.writeNR = false,
    this.notify = false,
    this.indicate = false,
  }) :
    uuid = uuid.toLowerCase(),
    value = value ?? Uint8List(0)
  ;

  Map<String, dynamic> toMap() {
    return {
      "uuid": uuid,
      "value": value,
      "read": read,
      "write": write,
      "writeNR": writeNR,
      "notify": notify,
      "indicate": indicate
    };
  }
}

File: android/src/main/kotlin/dev/steenbakker/flutter_ble_peripheral/models/CharacteristicDescription.kt
package dev.steenbakker.flutter_ble_peripheral.models

import android.bluetooth.BluetoothGattCharacteristic
import java.util.UUID

class CharacteristicDescription(
    uuid: String,
    val value: ByteArray,
    val read: Boolean,
    val write: Boolean,
    val writeNR: Boolean,
    val notify: Boolean,
    val indicate: Boolean,
) {
    val uuid = UUID.fromString(uuid)

    constructor(properties: Map<String, Any>) : this(
        properties["uuid"] as String,
        properties["value"] as ByteArray,
        properties["read"] as Boolean,
        properties["write"] as Boolean,
        properties["writeNR"] as Boolean,
        properties["notify"] as Boolean,
        properties["indicate"] as Boolean
    )

    fun properties() : Int {
        var ans: Int = 0
        if (read) ans = ans or BluetoothGattCharacteristic.PROPERTY_READ
        if (write) ans = ans or BluetoothGattCharacteristic.PROPERTY_WRITE
        if (writeNR) ans = ans or BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE
        if (notify) ans = ans or BluetoothGattCharacteristic.PROPERTY_NOTIFY
        if (indicate) ans = ans or BluetoothGattCharacteristic.PROPERTY_INDICATE
        return ans
    }

    fun permissions() : Int {
        var ans: Int = 0
        if (read) ans = ans or BluetoothGattCharacteristic.PERMISSION_READ
        if (write || writeNR) ans = ans or BluetoothGattCharacteristic.PERMISSION_WRITE
        return ans
    }
}